src/AggregateFunctions/AggregateFunctionDeepMergeJSON.cpp

```diff
@@ -0,0 +1,390 @@
+}
+
+/// Helper to validate paths count
+void validatePathsCount(size_t paths_count)
+{
+    if (paths_count > MAX_JSON_MERGE_PATHS)
+        throw Exception(
+            ErrorCodes::TOO_LARGE_ARRAY_SIZE, "Too many paths in JSON merge: {} (maximum: {})", paths_count, MAX_JSON_MERGE_PATHS);
+}
+
+/// Helper to validate total size
+void validateTotalSize(size_t total_size)
+{
+    if (total_size > MAX_JSON_MERGE_TOTAL_SIZE)
+        throw Exception(
+            ErrorCodes::TOO_LARGE_STRING_SIZE,
+            "JSON merge state size too large: {} bytes (maximum: {} bytes)",
+            total_size,
+            MAX_JSON_MERGE_TOTAL_SIZE);
+}
+}
+
+bool DeepMergeJSONAggregateData::isObjectPath(const StringRef & path) const
+{
+    auto it = paths.upper_bound(path);
+    return it != paths.end() && it->first.size > path.size && memcmp(it->first.data, path.data, path.size) == 0
+        && it->first.data[path.size] == '.';
+}
+
+bool DeepMergeJSONAggregateData::isUnsetMarker(const Field & value)
+{
+    if (value.getType() != Field::Types::Object)
+        return false;
+
+    const auto & obj = value.safeGet<Object>();
+    return obj.size() == 1 && obj.contains(UNSET_KEY) && obj.at(UNSET_KEY).getType() == Field::Types::Bool
+        && obj.at(UNSET_KEY).safeGet<bool>();
+}
+
+bool DeepMergeJSONAggregateData::handleDeletion(const StringRef & target_path, size_t order, Arena * arena)
+{
+    auto it = paths.find(target_path);
+
+    if (it != paths.end() && order <= it->second.row_order)
+        return false; // Skip if not newer
+
+    if (it == paths.end())
+    {
+        // Need to intern the string for new deletion marker
+        char * data = arena->alloc(target_path.size);
+        memcpy(data, target_path.data, target_path.size);
+        paths[StringRef(data, target_path.size)] = PathData{Field(), order, true};
+    }
+    else
+    {
+        it->second.value = Field();
+        it->second.row_order = order;
+        it->second.is_deleted = true;
+    }
+
+    removeChildPaths(target_path);
+    return true;
+}
+
+void DeepMergeJSONAggregateData::addPath(const StringRef & path, const Field & value, size_t order, Arena * arena)
+{
+    /// Check for ".$unset" suffix deletion
+    std::string path_str = path.toString();
+    std::string unset_suffix = std::string(".") + UNSET_KEY;
+    if (path_str.size() > unset_suffix.size() && path_str.substr(path_str.size() - unset_suffix.size()) == unset_suffix
+        && value.getType() == Field::Types::Bool && value.safeGet<bool>())
+    {
+        std::string target_path = path_str.substr(0, path_str.size() - unset_suffix.size());
+        handleDeletion(StringRef(target_path), order, arena);
+        return;
+    }
+
+    /// Check for deletion marker (backward compatibility)
+    if (isUnsetMarker(value))
+    {
+        handleDeletion(path, order, arena);
+        return;
+    }
```

Comment:
What do you mean by `backward compatibility` here? How is it possible to get an `Object` inside `value`?

Reply:
i was prototyping something earlier, i removed this.